src/TraceEvent/EventPipe/EventPipeEventSource.cs

```diff
@@ -34,21 +40,47 @@ public EventPipeEventSource(string fileName)
             cpuSpeedMHz = 10;
 
             _deserializer = new Deserializer(new PinnedStreamReader(fileName, 0x20000), fileName);
+
+#if SUPPORT_V1_V2
+            // This is only here for V2 and V1.  V3+ should use the name EventTrace, it can be removed when we drop support.
             _deserializer.RegisterFactory("Microsoft.DotNet.Runtime.EventPipeFile", delegate { return this; });
-        
-            var entryObj = _deserializer.GetEntryObject(); // this call invokes FromStream and reads header data
+#endif
+            _deserializer.RegisterFactory("!EventTrace.", delegate { return this; });
```

Comment:
Why do the names of the types begin with ! and end with .?     If it is just to avoid collision, we should not worry about this, as the namespace is just THIS FILE and no more.   I would prefer the names 'EventTrace' and 'EventBlock' (I could even be talked into simply 'Trace' and 'EventBlock' as these are the names you would use if you had to come up with names that need only be unique in this file.   (We may end up with other kinds of block, and frankly 'Block' is not descriptive enough on its own'.

Reply:
I am not very proud of this, but..

Every object is serialized as:

```xml
<BeginObject> size is 1
	<BeginObject> size is 2
		<NullReference> size is 3
		<objectVersion> size is 7
		<minReaderVersion> size is 11
		<stringLength> size is 15
		<string> size is 15 + stringLength
	</EndObject> size is 16 + stringLength
	<Content /> <-- this must be aligned, so stringLength has to be aligned too ;)
</EndObject>
```

So I decided to enforce a rule that type name length must be a multiplication of 4. This + padding the content ensures that everything is aligned. I know that it sounds stupid, but I did not have better idea. The other idea for names I had was EventBlock**4B**

@vancem is that acceptable?

Reply:
What you described is SUPER subtle and fragile (any change will break alignment).   You don't want one piece code (what goes in the headers) to be responsible for a subtle invariant needed later on (the alignment of a block).   

In general, whenever you write a piece of code, you should be thinking about some programmer (who might by you), a year later trying to figure out your code, and WANTING TO LEARN AS LITTLE AS POSSIBLE about it to get his job done.   Thus you want everything to be 'obvious' and 'explicit' as possible.   Being 'smart' is a BAD idea unless there is a compelling reason to (e.g. it improves perf in a non-trivial way, and then you have to pay that 'smartness' by documenting it)    Sometimes we fall short, but this is the goal.   

Given that you added a new 'Padding' tag, I figured what you did was add these to the 'tail' of the object before.   But thinking about it, I don't like that either, because it places the burden of alignment on the PREVISOUS object, and it is also pretty subtle (you need to know that there will be an end-object byte afterward and adjust for that). 

All of that is too subtle.   The 'obvious' thing to do is to 'align when you need it' INSIDE objects.   Thus where we need alignment is INSIDE the EventBlock.   Thus after emitting the block size, the writer will simply emit 0 to 3 bytes until the output pointer is 4 byte aligned (this counts as part of the payload).  Similarly,  the reader, after reading the block size will read bytes until its input pointer is 4 byte aligned.      This is just a few lines of code, and makes alignment very clear and explicit (frankly you don't need asserts, since you FORCED the alignment at that point.   

Notice this approach is robust (no matter what is happening OUTSIDE this object, you get the alignment you want.  It is LOCAL (only code associated with EventBlock needs to care, and it is EXPLICIT (there is code (near where you would care), that has the 'job' of providing alignment and programmers can see it (it is not a subtle ramification of something that seems to have another purpose).   

I would like us to use this approach (or something equally 'obvious') and then you can rip out the subtle names ! and . in the names).   

Seems like you also don't need to change IFastSerializable either.